### PR TITLE
Energy profiling tools: Add Daemon class for periodic task execution

### PR DESCRIPTION
**Pull request by ethan-puyaubreau**
This PR introduces a generic Daemon class for background task execution in the energy profiler infrastructure:

- Thread-safe periodic execution with configurable intervals
- Clean start/stop lifecycle management
- Reusable foundation for background monitoring tasks

The daemon system will enable energy providers to perform periodic power sampling at specified intervals, supporting the energy measurement capabilities planned in #301.

### CODE REVIEW

File: profiling/energy-profiler/daemon/daemon.hpp [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, int interval_ms)
+      : interval_(interval_ms), func_(func){};
+
+  void start();
+  void run();
+  void stop();
+  bool is_running() const { return running_; }
+  std::thread& get_thread() { return thread_; }
```
Comment (JBludau) [inline profiling/energy-profiler/daemon/daemon.hpp]: do we actually want this to be modifiable?

**Review by JBludau — APPROVED** [review]
Description and title could be updated, but the rest is ok for now

**Comment by JBludau** [comment]
If we go to c++20 we can use jthread for thread safety, but that is for kokkos 5

File: profiling/energy-profiler/daemon.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+#pragma once
+
+#include <functional>
+#include <thread>
+#include <chrono>
+
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, int interval_ms)
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]: Should we pass in a `std::chrono::duration` directly, instead of an `int`? What about the type `int`; e.g. what is the default arithmetic type that `std::chrono::milliseconds` encodes?

File: profiling/energy-profiler/daemon.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      : interval_(interval_ms), func_(func){};
+
+  void start();
+  void run();
+  void stop();
+  bool is_running() const { return running_; }
+  std::thread& get_thread() { return thread_; }
+
+ private:
+  std::chrono::milliseconds interval_;
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]: Should this be chosen as the smallest possible tick period? Thinking that "milliseconds" may be quite long as a sampling interval.

File: profiling/energy-profiler/daemon.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+  bool running_{false};
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp:38]: Is there an issue of thread safety?

File: profiling/energy-profiler/daemon.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#pragma once
+
+#include <functional>
+#include <thread>
+#include <chrono>
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]: Best to order the includes alphabetically.

File: profiling/energy-profiler/daemon.hpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+// Under the terms of Contract DE-NA0003525 with NTESS,
+// the U.S. Government retains certain rights in this software.
+//
+// Part of Kokkos, under the Apache License v2.0 with LLVM Exceptions.
+// See https://kokkos.org/LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#pragma once
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]: We should probably agree at the level of `kokkos-tools` on using consistently either preprocessor conditionals or this pragma to protect our includes.

File: profiling/energy-profiler/daemon.cpp [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+#include "daemon.hpp"
+#include <stdexcept>
+#include <thread>
+
+void Daemon::start() {
+  if (!running_) {
+    running_ = true;
+    thread_  = std::thread(&Daemon::run, this);
+  } else {
+    throw std::runtime_error("Daemon already started");
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.cpp]: We may want to think about how to be consistent in handling errors. In your tool, it appears you're sometimes using exceptions, and sometimes using a boolean to handle errors. I'm not sure if there's a policy at the `kokkos-tools` level to avoid exceptions. Probably best to check e.g. with @JBludau.

File: profiling/energy-profiler/daemon.cpp [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  }
+}
+
+void Daemon::run() {
+  while (running_) {
+    auto next_run = std::chrono::high_resolution_clock::now() + interval_;
+    func_();
+    std::this_thread::sleep_until(next_run);
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.cpp:40]: Just wondering about use cases where e.g. the time that the function takes is longer than the sampling interval.

**Review by maartenarnst — COMMENTED** [review]
This PR would also benefit from tests and an example.

File: profiling/energy-profiler/daemon.hpp (hunk 2 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: yes, that is a good idea.

File: profiling/energy-profiler/daemon.hpp (hunk 3 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: hmm ... we could use the smallest to make it as general as possible. Nevertheless, the question that arises is if there is anything useful below the ms threshold. Do you have an example in mind?

File: profiling/energy-profiler/daemon.hpp (hunk 4 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp:38]: yes. As stated in a previous comment, we could just switch to jthread and https://en.cppreference.com/w/cpp/thread/jthread/request_stop.html once we have c++20

File: profiling/energy-profiler/daemon.hpp (hunk 6 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: I agree that consistency would be a good thing to aim for. There are a lot more things in Tools that could be made consistent though.

File: profiling/energy-profiler/daemon.cpp (hunk 7 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp]: I would even say we don't want to throw when the daemon is already running. This is not an exception but should just do nothing if the daemon is started.

File: profiling/energy-profiler/daemon.cpp (hunk 8 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:40]: We never ment this to be able to run a function at an interval lower than its own execution time. That would be a much more complex daemon.

Comment (maartenarnst) [reply profiling/energy-profiler/daemon.cpp:40]: OK. Just thinking that in terms of strategy for the PR, it may make sense to clarify whether the PR is going for a general deamon, or for a deamon that works for the energy use case. Perhaps focusing on the energy use case would be helpful because it would give a good framework to think about issues like the choice of the tick for the duration, or this issue of the sampling frequency vs the duration of the API call.

If you narrow the focus to a daemon specific to the energy use case, the daemon could be placed into the `KokkosTools::EnergyProfiler` namespace for now.

That being said, I do think it would make sense to think about what happens when the API call takes longer than the sampling interval. Perhaps not systematically. But e.g. in the case where for whatever random reason, the API call happens to take a bit longer than expected.

Comment (vlkale) [reply profiling/energy-profiler/daemon.cpp:40]: First, let me preface by saying this work by @ethan-puyaubreau has been valuable in terms of capabilities for energy profiling and monitoring as well as ideas for software engineering that apply to other tools in Kokkos Tools.

With that, to hone in on your bigger picture question on a generic daemon class:

Assuming I have understood your use of the term Daemon as a real-time independent monitoring service, a key reason for my Kokkos Tools PR #291 for the Kokkos Tools LDMS connector is enabling this daemon-like functionality via a production-level performance monitoring software tool, called LDMS (Lightweight Distributed Monitoring Service). The motivation for it is that the Kokkos Tools LDMS connector can eventually integrated with - or  more specifically be a parent tools library to - any other Kokkos Tools child/base library, e.g., nvtx-connector, variorium-connector. A variant of this LDMS connector had been used for the last couple years in production - it has been closed-source until this PR #291. (By the way, PR #291 is ready for review - feel free to provide suggestions for improvements and/or let me know if you have questions on it.)

File: profiling/energy-profiler/daemon.hpp (hunk 2 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: done in https://github.com/kokkos/kokkos-tools/pull/300/commits/646bf35d0b4ec4935e3654833670bec8edf22464

File: profiling/energy-profiler/daemon.hpp (hunk 3 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: changed in https://github.com/kokkos/kokkos-tools/pull/300/commits/646bf35d0b4ec4935e3654833670bec8edf22464

File: profiling/energy-profiler/daemon.hpp (hunk 5 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: done in https://github.com/kokkos/kokkos-tools/pull/300/commits/626f77780fe53451883f1e0bc193d987915e99ea

File: profiling/energy-profiler/daemon.hpp (hunk 6 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp]: switched to classic ifdef in https://github.com/kokkos/kokkos-tools/pull/300/commits/67a80309e4e776863c24aebcf22b9bec5f023559

File: profiling/energy-profiler/daemon.cpp (hunk 7 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp]: removed the exceptions in https://github.com/kokkos/kokkos-tools/pull/300/commits/45028511a5faeca42e030898aa5975a2dd5a1360

File: profiling/energy-profiler/daemon.cpp (hunk 8 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:40]: I moved it into the `KokkosTools::EnergyProfiler` namespace in https://github.com/kokkos/kokkos-tools/pull/300/commits/f01b82e2ce0d061c9e38e98c66259f537c785302

File: profiling/energy-profiler/daemon.cpp [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//@HEADER
+
+#include "daemon.hpp"
+#include <stdexcept>
+#include <thread>
+
+namespace KokkosTools {
+namespace EnergyProfiler {
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.cpp]:
```suggestion
namespace KokkosTools::EnergyProfiler {
```

File: profiling/energy-profiler/daemon.hpp [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+
+#ifndef KOKKOSP_ENERGY_PROFILER_DAEMON_HPP
+#define KOKKOSP_ENERGY_PROFILER_DAEMON_HPP
+
+#include <chrono>
+#include <functional>
+#include <thread>
+
+namespace KokkosTools {
+namespace EnergyProfiler {
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]:
```suggestion
namespace KokkosTools::EnergyProfiler {
```

File: profiling/energy-profiler/daemon.hpp [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, std::chrono::duration interval)
+      : interval_(interval), func_(func){};
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]:
```suggestion
      : interval_(interval), func_(std::move(func)){};
```

File: profiling/energy-profiler/daemon.hpp [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+#include <chrono>
+#include <functional>
+#include <thread>
+
+namespace KokkosTools {
+namespace EnergyProfiler {
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, std::chrono::duration interval)
```
Comment (maartenarnst) [inline profiling/energy-profiler/daemon.hpp]:
```suggestion
  Daemon(std::function<void()> func, const std::chrono::duration& interval)
```

**Comment by maartenarnst** [comment]
Hey @JBludau. I've just gone over the changes. You'll find a few suggestions. Sill thinking that it would be good to add a small test. Perhaps there could be a test with a callback function that just waits for a given duration. The test could use durations that would be typical for the use case of the energy profiler.

**Comment by JBludau** [comment]
> Hey @JBludau. I've just gone over the changes. You'll find a few suggestions. Sill thinking that it would be good to add a small test. Perhaps there could be a test with a callback function that just waits for a given duration. The test could use durations that would be typical for the use case of the energy profiler.

Yeah, I asked @ethan-puyaubreau if he could add his test.

File: profiling/energy-profiler/daemon.hpp [hunk 13]
```diff
@@ -0,0 +1,44 @@
...
+namespace KokkosTools::EnergyProfiler {
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, const std::chrono::duration& interval)
+      : interval_(interval), func_(std::move(func)){};
+
+  void start();
+  void stop();
+  bool is_running() const { return running_; }
+  std::thread& get_thread() { return thread_; }
```
Comment (aprokop) [inline profiling/energy-profiler/daemon.hpp]:
```suggestion
  auto& get_thread() { return thread_; }
```

File: profiling/energy-profiler/daemon.cpp [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+//
+//@HEADER
+
+#include "daemon.hpp"
+#include <stdexcept>
+#include <thread>
+
+namespace KokkosTools::EnergyProfiler {
+void Daemon::start() {
+  if (!running_) {
```
Comment (aprokop) [inline profiling/energy-profiler/daemon.cpp:23]: What are the semantics of `start()` if it's already running? Should it break? Should it be restarted?

File: profiling/energy-profiler/daemon.cpp [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    running_ = true;
+    thread_  = std::thread(&Daemon::run, this);
+  }
+}
+
+void Daemon::stop() {
+  if (running_) {
```
Comment (aprokop) [inline profiling/energy-profiler/daemon.cpp:30]: What are the semantics here? Should we require that it's been running?

**Review by aprokop — COMMENTED** [review]
This looks good to me. I only have a couple question.
We need this capability as soon as we can for one of my projects.

File: profiling/energy-profiler/daemon.cpp (hunk 14 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:23]: I think it should just do nothing if it is already running. Multiple start calls should be fine since in our current use case the global system time is recorded. If someone wants to use it differently I would opt we add a `restart()` functionality.

File: profiling/energy-profiler/daemon.cpp (hunk 15 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:30]: I mean, it is not threadsafe, so we are limited with the current impl. But if we care we can either go to jthread or at least use atomics for the status

File: profiling/energy-profiler/daemon.cpp (hunk 14 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:23]: but as described below ... this is not threadsafe anyway

**Review by aprokop — APPROVED** [review]

File: profiling/energy-profiler/daemon.cpp (hunk 15 above)
Comment (vlkale) [reply profiling/energy-profiler/daemon.cpp:30]: FWIW, I think C++20 `std::jthread` is a great idea here. I have a slight inclination to use C++ atomics for now since it seems more straightforward - I am wondering how portable/available jthreads would be on various computing systems currently?

File: profiling/energy-profiler/daemon.cpp (hunk 8 above)
Comment (vlkale) [reply profiling/energy-profiler/daemon.cpp:40]: Thanks!

**Comment by vlkale** [comment]
Hi all, I was wondering about the status on this. I think this is a good PR that would be valuable for a lot of profiling data beyond energy information, which is relevant to #291 - albeit I would say the energy profiling use case is the one that is probably of most value currently to production Kokkos applications. 

Do you know the rough timeline of this PR by chance? I know that @ethan-puyaubreau is done with his internship, so there may be timing constraints from his end. 

I think this PR is on the right track overall.

**Comment by JBludau** [comment]
The main open thing is a desire to have an example for the daemon. So far I was not able to come up with a good one.
Personally I think it is fine without an example as the energy profiling pr's that depend on this are basically going to be example usages of this.

**Comment by vlkale** [comment]
> The main open thing is a desire to have an example for the daemon. So far I was not able to come up with a good one. Personally I think it is fine without an example as the energy profiling pr's that depend on this are basically going to be example usages of this.

Yes, I definitely agree just considering the energy profiling use case already shows an example usage. It would certainly help to have a small example since it can make clear assumptions being made on its use, but I think it doesn't need to be a blocker. I do think some of the error handling and/or tests for corner cases (e.g., interval being longer than time for kernel) that are discussed by @maartenarnst and you in this PR are important to think about.  

Except for making sure no critical error checks are missing, I think the code logic of this PR makes sense to me. 

Doing some sort of beta release of this Daemon class, especially so, e.g., @aprokop along with #291 can use this would be good to get early feedback on real-world production use cases, as that could directly provide a lot of insight for improvements. Maybe it could be merged in as a beta feature early next year? 

Just my thoughts on this now - I will look and see if I can make a few more improvements to the code and error checks in the PR myself till then.

File: profiling/energy-profiler/daemon.cpp (hunk 7 above)
Comment (vlkale) [reply profiling/energy-profiler/daemon.cpp]: Yeah, exceptions aren't really used for Kokkos Tools libraries - and I think not throwing here is the right choice.

File: profiling/energy-profiler/daemon.hpp (hunk 3 above)
Comment (vlkale) [reply profiling/energy-profiler/daemon.hpp]: > changed in [646bf35](https://github.com/kokkos/kokkos-tools/commit/646bf35d0b4ec4935e3654833670bec8edf22464)

Sounds good though at least for interactions with performance monitoring systems (including for #291 for LDMS), I don't think there is any benefit (and it may actually be detrimental) to go below the millisecond level in production scenarios.

File: profiling/energy-profiler/daemon.hpp [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+#define KOKKOSP_ENERGY_PROFILER_DAEMON_HPP
+
+#include <chrono>
+#include <functional>
+#include <thread>
+
+namespace KokkosTools::EnergyProfiler {
+class Daemon {
+ public:
+  Daemon(std::function<void()> func, const std::chrono::duration& interval)
```
Comment (vlkale) [inline profiling/energy-profiler/daemon.hpp:27]: One quick note is I don't see a default value for the time interval of the Daemon but maybe you could have one. If you do, I think a reasonable one would be 1 second. This would at least makes sense from the point of view of #291. Maybe there are other use cases and situations where you want it to be shorter (or longer).

**Review by vlkale — COMMENTED** [review]
LGTM overall, just a suggestion to consider a default interval

File: profiling/energy-profiler/daemon.hpp (hunk 16 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.hpp:27]: Hmm, I think the user explicitly specifying a duration is a good practice here. It is a core setting of the daemon that should be thought about explicitly.

Comment (vlkale) [reply profiling/energy-profiler/daemon.hpp:27]: Ah, got it - yes, I agree. We want this Daemon class to be generic beyond the energy profiling use case and/or LDMS use case, and it should indeed be thought out by users explicitly based on the scenario.

**Comment by JBludau** [comment]
@maartenarnst do you want us to investigate the use case of a longer return time than the frequency the daemon is given? If so, do you have a specific case in mind that would need this?

File: profiling/energy-profiler/daemon.cpp (hunk 15 above)
Comment (JBludau) [reply profiling/energy-profiler/daemon.cpp:30]: Currently the tools don't have a minimum requirement for 20, right? Since most of the other tools are not threadsafe, I would say that is a larger discussion.
But if we want it to be `std::jthread` now, we can update. I am happy for any contribution

**Comment by JBludau** [comment]
@vlkale I think we can merge this. Maarten did not reply and we still can look at making the daemon able to deal with runtimes longer than its frequency later if we ever have a usecase for this

**Comment by vlkale** [comment]
> @vlkale I think we can merge this. Maarten did not reply and we still can look at making the daemon able to deal with runtimes longer than its frequency later if we ever have a usecase for this

@JBludau I am good with merging it. Yes, allowing for runtimes longer than its frequency would be good but it's not needed right now for immediate/priority use cases.

**Comment by JBludau** [comment]
@vlkale  could you merge this?

**Comment by vlkale** [comment]
@dalg24 @crtrott Can you please review and merge if it looks good to you?

This is a simple but important addition to Kokkos Tools for integrating profiling of system-level metrics like energy consumption.

**Review by dalg24 — APPROVED** [review]
